psutil/arch/solaris/process_as_utils.c

```diff
@@ -0,0 +1,302 @@
+
+#if !defined(_LP64) && _FILE_OFFSET_BITS == 64
+#  undef _FILE_OFFSET_BITS
+#  undef _LARGEFILE64_SOURCE
+#endif
+
+#include <Python.h>
+
+#include <sys/types.h>
+#include <sys/procfs.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include "process_as_utils.h"
+
+static int
+open_address_space(pid_t pid, const char *procfs_path) {
+    int fd;
+    char proc_path[PATH_MAX];
+
+    sprintf(proc_path, "%s/%i/as", procfs_path, pid);
+    fd = open(proc_path, O_RDONLY);
+    if (fd < 0)
+        PyErr_SetFromErrno(PyExc_OSError);
+
+    return fd;
+}
+
+static int
+read_offt(int fd, off_t offset, char *buf, size_t buf_size) {
+    size_t to_read = buf_size;
+    size_t stored  = 0;
+
+    while (to_read) {
+        int r = pread(fd, buf + stored, to_read, offset + stored);
+        if (r < 0) {
+            PyErr_SetFromErrno(PyExc_OSError);
+            return -errno;
+        }
+
+        to_read -= r;
+        stored += r;
+    }
+
+    return stored;
+}
+
+#define STRING_SEARCH_BUF_SIZE 512
+
+static char *
+read_cstring_offt(int fd, off_t offset) {
+    int r;
+    int i;
+    off_t end;
+    size_t len;
+    char buf[STRING_SEARCH_BUF_SIZE];
+    char *result = NULL;
+
+    if (lseek(fd, offset, SEEK_SET) == (off_t)-1) {
+        PyErr_SetFromErrno(PyExc_OSError);
+        return NULL;
+    }
+
+    // Search end of string
+    end = offset;
+    for (;; end += STRING_SEARCH_BUF_SIZE) {
+        r = read(fd, buf, sizeof(buf));
+
+        if (r == -1) {
+            PyErr_SetFromErrno(PyExc_OSError);
+            return NULL;
+        }
+        else if (r == 0) {
+            break;
+        }
+        else {
+            for (i=0; i<r; i++) {
+                if (! buf[i]) {
+                    end += i;
+                    goto lbEnd;
+                }
+            }
+        }
+    }
+
+  lbEnd:
+    len = end - offset;
+
+    result = malloc(len+1);
+    if (! result)
+        return (char *) PyErr_NoMemory();
+
+    if (len) {
+        if (read_offt(fd, offset, result, len) < 0) {
+            free(result);
+            return NULL;
+        }
+    }
+
+    result[len] = '\00';
+    return result;
+}
+
+static char **
+read_cstrings_block(int fd, off_t offset, size_t ptr_size, size_t count) {
+    char **result = NULL;
+    char *pblock;
+    size_t pblock_size;
+    int i;
+
+    assert(ptr_size == 4 || ptr_size == 8);
+
+    if (!count)
+        goto lbExit;
+
+    pblock_size = ptr_size * count;
+
+    pblock = malloc(pblock_size);
+    if (! pblock)
+        return (char **) PyErr_NoMemory();
+
+    if (read_offt(fd, offset, pblock, pblock_size) != pblock_size) {
+        goto lbPBlockError;
+    }
+
+    result = (char **) calloc(count, sizeof(char *));
+    if (! result) {
+        PyErr_NoMemory();
+        goto lbPBlockError;
+    }
+
+    for (i=0; i<count; i++)
+        result[i] = read_cstring_offt(
+            fd, (ptr_size == 4?
+                ((uint32_t *) pblock)[i]:
+                ((uint64_t *) pblock)[i]));
```

Comment:
I don't like this, also because this function fully implements error handling, but if the caller does not check the return value then what's the point? Did you see this function fail?